tests/tools.py

```diff
@@ -543,3 +544,10 @@ def wrapper(self, *args, **kwargs):
     return decorator
 
 
+def clear_singleton_instances(cls):
```

Comment:
Can you add a comment explaining why this was introduced? For example, "some tests like XY use singleton instances of ProtocolUtil because ... ", just so it's easier to follow when reading the code.

Reply:
I feel adding more comments here explaining what the function does might make it extra verbose given the fact that I've already added comments everywhere ProtocolUtil was cleared. (Additionally I think the function name is pretty self-explanatory). I'll skip adding more comments for now but we can always add it later if we feel the need for it :)

Reply:
Fair enough :)